feat(iota-core): update neural network model server from congestion tracker module

salaheldinsoliman opened the pull request:
# Description of change

This PR adds update to the neural network third party server via a function `inform_model`. `inform_model` is called at the end of `process_checkpoint_effects`.
Any logic regarding the update itself is placed in `model_updater.rs`, that is not to clutter congestion_tracker with un-related logic

## Links to any relevant issues

Be sure to reference any related issues by adding `fixes #(issue)`.

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

salaheldinsoliman committed:
cleanup: update nn model server from congestion tracker

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

salaheldinsoliman committed:
cleanup: update nn model server from congestion tracker

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

salaheldinsoliman committed:
update docker-compose to contain preditcor service

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

vekkiokonio commented on crates/iota-core/src/congestion_tracker.rs:
```diff
@@ -139,9 +143,109 @@ pub struct CongestionTracker {
+    ) {
+        // 1) Build touched set
+        let mut touched: std::collections::HashSet<ObjectID> = std::collections::HashSet::new();
+        for tx in congestion_txs_data {
```
Here you don't differentiate between write shared objects, read shared objects and owned objects. That means that read-only or owned objects would affect the prediction, which shouldn't be the case

vekkiokonio replied:
For instance, check the filter at line 327

vekkiokonio replied:
@salaheldinsoliman I think my comment was wrong here: you are using the same `congestion_txs_data` and `cleared_txs_data` from `process_checkpoint_effect()` so there is actually no need to include the mutate filter for congested objects -- in fact, we can only have congestion for write shared objects. I suggest to revert https://github.com/iotaledger/iota/pull/8993/commits/ddb1243d1d1071b25ca705766c0734eea8f54065

vekkiokonio commented on crates/iota-core/src/congestion_tracker.rs:
```diff
@@ -139,9 +143,109 @@ pub struct CongestionTracker {
+                        latest_congestion_time: Some(info.latest_congestion_time),
+                        highest_congestion_gas_price: info.highest_congestion_gas_price,
+                        latest_clearing_time: info.latest_clearing_time,
+                        lowest_clearing_gas_price: info.lowest_clearing_gas_price.unwrap_or(0),
```
```suggestion
                        lowest_clearing_gas_price: info.lowest_clearing_gas_price.unwrap_or(self.reference_gas_price),
```

vekkiokonio commented on crates/iota-core/src/congestion_tracker.rs:
```diff
@@ -139,9 +143,109 @@ pub struct CongestionTracker {
+            }
+        }
+
+        // 3) Build per-checkpoint stats
```
Both 2) and 3) are "per-checkpoint stats". The difference is that 2) is retrieved directly from congestion_info, while 3) is computed in this function. Stats in 2) provide richer information as they also include gas prices. I'm wondering if stats in 3) are really needed and if they actually contributed to improve the NN prediction.

salaheldinsoliman replied:
Good point :+1: 
Step (2) is a post‑checkpoint snapshot from the cache (latest_* times, gas‑price extremes, hotness), 
while step (3) adds per‑checkpoint intensity: how many congested/clearing txs touched the object. The booleans in (3) are redundant (derivable from (2) by comparing latest_*_time to this checkpoint), but the counts are not and provide extra signal for the predictor. 
I think we should keep the counts for now, WDYT?

vekkiokonio replied:
Sure, there is no harm to keep them

vekkiokonio commented on crates/iota-core/src/congestion_tracker.rs:
```diff
@@ -139,9 +143,109 @@ pub struct CongestionTracker {
+            congestion_txs_data.len() + clearing_txs_data.len(),
+        );
+        let mut per_obj_min_clearing: HashMap<ObjectID, u64> = HashMap::new();
+        for tx in congestion_txs_data {
```
You introduce several for loops over the same data structures (`clearing_txs_data` and `congestion_txs_data`). I think you can combine 1), 3) and 5) into a single loop.

salaheldinsoliman replied:
Thanks, 
I replaced the separate loops for touched(only mutated), stats, and training items with one loop over `congestion_txs_data` + `clearing_txs_data`

vekkiokonio commented on crates/iota-core/src/congestion_tracker.rs:
```diff
@@ -139,9 +143,109 @@ pub struct CongestionTracker {
+                gas_price_feedback: None,
+                touched_objects: tx.objects.clone(),
+            });
+            for oid in &tx.objects {
```
Your `per_obj_min_clearing` vector corresponds to `get_prediction_suggested_gas_price` computed for all objects in the checkpoint. However, your code doesn't handle the rare but possible case where no clearing data are present and all transactions are canceled for a given checkpoint

salaheldinsoliman replied:
nice catch, what I did now is basically:
A second accumulator that tracks, per object, the highest congestion requirement seen in the checkpoint (using gas_price_feedback if present, otherwise the gas price,). 
After the single pass over transactions, I merge the two into a per_obj_required_in_cp map: it prefers the lowest clearing price when available, and otherwise falls back to that highest congestion value.

vekkiokonio replied:
Here, I would suggest to reuse the existing congestion tracker logic. Specifically, I would suggest to directly use the `get_suggested_gas_price_for_objects()` function which already does what you want. I would remove `per_obj_min_clearing` and `per_obj_max_congestion` and use something like:

```
let price = self.get_suggested_gas_price_for_objects(std::iter::once(*oid));
```

to store data into `per_obj_required_in_cp`.

salaheldinsoliman committed:
add congestion audit to .gitignore

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

vekkiokonio commented on crates/iota-core/src/model_updater.rs:
```diff
@@ -0,0 +1,480 @@
+use iota_types::base_types::ObjectID;
+
+/// Default base URL for the predictor service.
+pub const DEFAULT_PREDICTOR_BASE_URL: &str = "http://predictor:9666";
```
Had to change this to `"http://0.0.0.0:9666"` to make it work

salaheldinsoliman replied:
This has to run in the docker compose setup to work with this URL, I will add a comment for clarification

salaheldinsoliman committed:
fix: only consider mutated shared objects

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

salaheldinsoliman committed:
fix: remove unneeded for loops

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

salaheldinsoliman committed:
fix: address no clearing txs case

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

salaheldinsoliman committed:
fix: clarify predictor url usage

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

salaheldinsoliman committed:
Merge branch 'protocol-research/ogd-predictor-experiments' into protocol-research/nn-prediction-model-server

salaheldinsoliman commented:
Changes summary to address the PR comments:
- Consider only mutated objects, not all objects
- Refactored multiple for loops into a single pass
- Addressed the case where no clearing Txs are in a checkpoint

vekkiokonio reviewed (approved):
LGTM

Please address the two comments I added to the PR and then it can be merged.

salaheldinsoliman committed:
Revert "fix: only consider mutated shared objects"

This reverts commit ddb1243d1d1071b25ca705766c0734eea8f54065.

salaheldinsoliman committed:
Merge branch 'protocol-research/nn-prediction-model-server' of github.com:iotaledger/iota into protocol-research/nn-prediction-model-server

salaheldinsoliman committed:
refactor: re-use get_suggested_gas_price() instead of custom logic

Signed-off-by: salaheldinsoliman <salaheldin_sameh@aucegypt.edu>

salaheldinsoliman commented:
Hi @vekkiokonio, I addressed the remaining two comments :+1:

vekkiokonio commented:
Please flag this as draft since it won't be merged any time soon
